wicketd/src/update_tracker.rs

```diff
@@ -591,57 +618,80 @@ impl UpdateDriver {
-                    elapsed,
-                },
-                ProgressEventKind::FormatProgress {
-                    attempt,
-                    path,
-                    percentage,
-                    elapsed,
-                } => UpdateStateKind::InstallinatorFormatProgress {
-                    attempt,
-                    path,
-                    percentage,
-                    elapsed,
-                },
-                ProgressEventKind::WriteProgress {
-                    attempt,
-                    kind,
-                    destination,
-                    written_bytes,
-                    total_bytes,
-                    elapsed,
-                } => UpdateStateKind::ArtifactWriteProgress {
-                    attempt,
-                    kind,
-                    destination: Some(destination),
-                    written_bytes,
-                    total_bytes,
-                    elapsed,
-                },
-            }
+        let progress_kind =
+            if let Some(event) = report.progress_events.drain(..).next() {
+                match event.kind {
+                    ProgressEventKind::DownloadProgress {
+                        attempt,
+                        kind,
+                        peer: _peer,
+                        downloaded_bytes,
+                        total_bytes,
+                        elapsed,
+                    } => UpdateStateKind::ArtifactDownloadProgress {
+                        attempt,
+                        kind,
+                        downloaded_bytes,
+                        total_bytes,
+                        elapsed,
+                    },
+                    ProgressEventKind::FormatProgress {
+                        attempt,
+                        path,
+                        percentage,
+                        elapsed,
+                    } => UpdateStateKind::InstallinatorFormatProgress {
+                        attempt,
+                        path,
+                        percentage,
+                        elapsed,
+                    },
+                    ProgressEventKind::WriteProgress {
+                        attempt,
+                        kind,
+                        destination,
+                        written_bytes,
+                        total_bytes,
+                        elapsed,
+                    } => UpdateStateKind::ArtifactWriteProgress {
+                        attempt,
+                        kind,
+                        destination: Some(destination),
+                        written_bytes,
+                        total_bytes,
+                        elapsed,
+                    },
+                }
+            } else {
+                UpdateStateKind::WaitingForProgress {
+                    component: "installinator".to_owned(),
+                }
+            };
+
+        if report.completion_events.is_empty() {
+            self.set_current_update_state(progress_kind);
```

Comment:
I don't think we need to do this. If the installinator sends both a progress and a completion message, it means that prior steps have completed and the installinator has moved to the next step.

Reply:
Sorry, not sure I follow. If we get a report with no completion events we still want to set the current progress, right?

Reply:
Yes, but if progress is available we should unconditionally set it.

Reply:
Oh! I see. :facepalm: Thanks, will fix momentarily.

Reply:
Actually, I think the code does this already! If the completion set is empty, it is set here; if the completion set is not empty, it is set (potentially repeatedly, but guaranteed at least once) in the iteration over the events.

Reply:
Ahh, I was looking for a `set_current_update_state` call in the other branch and totally missed that.